ml-pipeline-persistenceagent fails a few times.

Kubeflow v0.4.0-rc.3
On GKE (with official doc instructions for CLI setup)

Everything sets up well, but the `ml-pipeline-persistenceagent` has failed 4x. On the fifth retry, the pod is up and running. Here's a log of one of its failures.

If this is nothing to worry about, feel free to just close this issue.

```
$ kubectl logs ml-pipeline-persistenceagent-9ff99498c-jbq44 -p
W0104 17:36:19.094203       1 client_config.go:552] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1072ecd]

goroutine 1 [running]:
github.com/kubeflow/pipelines/backend/src/common/util.WaitForAPIAvailable.func1(0xc0002ff560, 0xc000302060)
	/go/src/github.com/kubeflow/pipelines/backend/src/common/util/service.go:38 +0x12d
github.com/kubeflow/pipelines/vendor/github.com/cenkalti/backoff.RetryNotify(0xc000405020, 0x14d9e60, 0xc0002ff560, 0x0, 0xc000269500, 0xc0002ff560)
	/go/src/github.com/kubeflow/pipelines/vendor/github.com/cenkalti/backoff/retry.go:37 +0xa2
github.com/kubeflow/pipelines/vendor/github.com/cenkalti/backoff.Retry(0xc000405020, 0x14d9e60, 0xc0002ff560, 0x2b, 0x70)
	/go/src/github.com/kubeflow/pipelines/vendor/github.com/cenkalti/backoff/retry.go:25 +0x48
github.com/kubeflow/pipelines/backend/src/common/util.WaitForAPIAvailable(0x1bf08eb000, 0x1371890, 0xd, 0xc000450ff0, 0x2b, 0xc000451020, 0x2b)
	/go/src/github.com/kubeflow/pipelines/backend/src/common/util/service.go:47 +0xb7
github.com/kubeflow/pipelines/backend/src/agent/persistence/client.NewPipelineClient(0xc00004440e, 0x8, 0x1bf08eb000, 0xdf8475800, 0x1371890, 0xd, 0x1370308, 0xb, 0x136817f, 0x4, ...)
	/go/src/github.com/kubeflow/pipelines/backend/src/agent/persistence/client/pipeline_client.go:60 +0x349
main.main()
	/go/src/github.com/kubeflow/pipelines/backend/src/agent/persistence/main.go:83 +0x287
```

--- a/backend/src/common/util/service.go
+++ b/backend/src/common/util/service.go
@@ -31,15 +31,17 @@ import (
 func WaitForAPIAvailable(initializeTimeout time.Duration, basePath string, apiAddress string) error {
 	var operation = func() error {
 		response, err := http.Get(fmt.Sprintf("http://%s%s/healthz", apiAddress, basePath))
-		if err == nil {
-			return nil
+		if err != nil {
+			return err
 		}
-		// we wait only on 503 service unavailable. Stop retry otherwise.
-		if response.StatusCode != 503 {
+
+		// If we get a 503 service unavailable, it's a non-retriable error.
+		if response.StatusCode == 503 {
 			return backoff.Permanent(errors.Wrapf(
 				err, "Waiting for ml pipeline API server failed with non retriable error."))
 		}
-		return err
+
+		return nil
 	}
 
 	b := backoff.NewExponentialBackOff()
@@ -49,7 +51,7 @@ func WaitForAPIAvailable(initializeTimeout time.Duration, basePath string, apiAd
 }
 
 func GetKubernetesClientFromClientConfig(clientConfig clientcmd.ClientConfig) (
-		*kubernetes.Clientset, *rest.Config, string, error) {
+	*kubernetes.Clientset, *rest.Config, string, error) {
 	// Get the clientConfig
 	config, err := clientConfig.ClientConfig()
 	if err != nil {

--- a/backend/test/test_utils.go
+++ b/backend/test/test_utils.go
@@ -44,20 +44,23 @@ var runIntegrationTests = flag.Bool("runIntegrationTests", false, "Whether to al
 func waitForReady(namespace string, initializeTimeout time.Duration) error {
 	var operation = func() error {
 		response, err := http.Get(fmt.Sprintf("http://ml-pipeline.%s.svc.cluster.local:8888/apis/v1beta1/healthz", namespace))
-		if err == nil {
-			return nil
+		if err != nil {
+			return err
 		}
-		// we wait only on 503 service unavailable. Stop retry otherwise.
-		if response.StatusCode != 503 {
-			return backoff.Permanent(errors.Wrapf(err, "Waiting for ml pipeline failed with non retriable error."))
+
+		// If we get a 503 service unavailable, it's a non-retriable error.
+		if response.StatusCode == 503 {
+			return backoff.Permanent(errors.Wrapf(
+				err, "Waiting for ml pipeline API server failed with non retriable error."))
 		}
-		return err
+
+		return nil
 	}
 
 	b := backoff.NewExponentialBackOff()
 	b.MaxElapsedTime = initializeTimeout
 	err := backoff.Retry(operation, b)
-	return errors.Wrapf(err, "Waiting for ml pipeline failed after all attempts.")
+	return errors.Wrapf(err, "Waiting for ml pipeline API server failed after all attempts.")
 }
 
 func getClientConfig(namespace string) clientcmd.ClientConfig {

